src/pudl_archiver/depositors/fsspec.py

```diff
@@ -0,0 +1,272 @@
+
+
+class Deposition(BaseModel):
+    """Represent an fsspec deposition."""
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    deposition_path: UPath
+    #: static list of files from deposition prior to any modifications
+    file_list: list[str]
+    version: str = "0.1"
+
+    @classmethod
+    def from_upath(cls, deposition_path: UPath):
+        """Construct deposition object from fsspec path to deposition."""
+        assert deposition_path.exists()
+        file_list = [
+            str(child.name)
+            for child in deposition_path.iterdir()
+            if child.name != deposition_path.name
+        ]
+        return cls(
+            deposition_path=deposition_path,
+            file_list=file_list,
+        )
+
+
+class FsspecAPIClient(DepositorAPIClient):
+    """Implement API for fsspec based depositors."""
+
+    base_path: str
+
+    @classmethod
+    async def initialize_client(
+        cls,
+        session: aiohttp.ClientSession,
+        sandbox: bool,
+    ) -> "FsspecAPIClient":
+        """Return initialized fsspec api client."""
+        return cls()
+
+    async def get_deposition(self, dataset_id: str):
+        """Get latest version of deposition associated with dataset_id."""
+        return Deposition.from_upath(
+            deposition_path=UPath(self.base_path) / dataset_id,
+        )
+
+    async def create_new_deposition(self, dataset_id: str):
+        """Prepare new deposition associated with dataset_id."""
+        deposition_path = UPath(self.base_path) / dataset_id
+        deposition_path.mkdir(parents=True, exist_ok=True)
+
+        return Deposition.from_upath(deposition_path=deposition_path)
+
+    def get_file(self, deposition: Deposition, filename: str) -> bytes:
+        """Download file from deposition."""
+        with (deposition.deposition_path / filename).open("rb") as f:
+            return f.read()
+
+
+class FsspecPublishedDeposition(PublishedDeposition):
+    """Represents published version of fsspec deposition."""
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    settings: RunSettings
+    deposition: Deposition
+    api_client: FsspecAPIClient
+    dataset_id: str
+
+    async def list_files(self):
+        """List files."""
+        return self.deposition.file_list
+
+    def get_deposition_link(self) -> str:
+        """Return link to deposition."""
+        return self.deposition.deposition_path.as_uri()
+
+    async def get_file(self, filename: str) -> bytes:
+        """Download file from deposition."""
+        return self.api_client.get_file(self.deposition, filename)
+
+    async def open_draft(self) -> "FsspecDraftDeposition":
+        """Open a new draft to make edits."""
+        return FsspecDraftDeposition(
+            deposition=self.deposition,
+            settings=self.settings,
+            api_client=self.api_client,
+            dataset_id=self.dataset_id,
+        )
+
+
+class FsspecDraftDeposition(DraftDeposition):
```

Comment:
Draft/published isn't a meaningful distinction right now since we haven't implemented any kind of versioning for the fsspec backend, but I think that's more a sign that this backend is still immature and we wouldn't want to use it for our main PUDL archives at this point.

Ideally, we will eventually expand this backend to support versioning, so there would be a published archive which is read only and will always point to a static version, and a draft which can be modified and eventually published.

Reply:
There's some logical notion of a deposition being "done" or not, which I think this captures, despite there not being a difference on disk right now.

I suppose we could rename `draft`/`published` as `readwrite`, `readonly` which might be clearer? 🤷